pkg/core/statesync/module.go

```diff
@@ -0,0 +1,469 @@
+	if err != nil {
+		return fmt.Errorf("failed to store state synchronisation point %d: %w", p, err)
+	}
+	s.syncStage = initialized
+	s.log.Info("try to sync state for the latest state synchronisation point",
+		zap.Uint32("point", p),
+		zap.Uint32("evaluated chain's blockHeight", currChainHeight))
+
+	// check headers sync state first
+	ltstHeaderHeight := s.bc.HeaderHeight()
+	if ltstHeaderHeight > p {
+		s.syncStage = headersSynced
+		s.log.Info("headers are in sync",
+			zap.Uint32("headerHeight", s.bc.HeaderHeight()))
+	}
+
+	// check blocks sync state
+	s.blockHeight = s.getLatestSavedBlock(p)
+	if s.blockHeight >= p {
+		s.syncStage |= blocksSynced
+		s.log.Info("blocks are in sync",
+			zap.Uint32("blockHeight", s.blockHeight))
+	}
+
+	// check MPT sync state
+	if s.blockHeight > p {
+		s.syncStage |= mptSynced
+		s.log.Info("MPT is in sync",
+			zap.Uint32("stateroot height", s.bc.GetStateModule().CurrentLocalHeight()))
+	} else if s.syncStage == headersSynced {
+		header, err := s.bc.GetHeader(s.bc.GetHeaderHash(int(p + 1)))
+		if err != nil {
+			return fmt.Errorf("failed to get header to initialize MPT billet: %w", err)
+		}
+		err = s.billet.Init(header.PrevStateRoot)
+		if err != nil {
+			return fmt.Errorf("failed to initialize MPT billet: %w", err)
+		}
+		s.log.Info("MPT billet initialized",
+			zap.Uint32("height", s.syncPoint),
+			zap.String("state root", header.PrevStateRoot.StringBE()))
+		pool := NewPool()
+		pool.Add(header.PrevStateRoot, []byte{})
+		err = s.billet.Traverse(func(n mpt.Node, _ []byte) bool {
+			nPaths, ok := pool.TryGet(n.Hash())
+			if !ok {
+				// if this situation occurs, then it's a bug in MPT pool or Traverse.
+				panic("failed to get MPT node from the pool")
+			}
+			pool.Remove(n.Hash())
+			childrenPaths := make(map[util.Uint256][][]byte)
+			for _, path := range nPaths {
+				nChildrenPaths := mpt.GetChildrenPaths(path, n)
+				for hash, paths := range nChildrenPaths {
+					childrenPaths[hash] = append(childrenPaths[hash], paths...) // it's OK to have duplicates, they'll be handled by mempool
+				}
+			}
+			pool.Update(nil, childrenPaths)
+			return false
+		}, true)
+		if err != nil {
+			return fmt.Errorf("failed to traverse MPT while initialization: %w", err)
+		}
+		s.mptpool.Update(nil, pool.GetAll())
+		if s.mptpool.Count() == 0 {
+			s.syncStage |= mptSynced
+			s.log.Info("MPT is in sync",
+				zap.Uint32("stateroot height", p))
+		} else {
+			unknownNodes := make(map[util.Uint256]bool)
+			for hash := range s.mptpool.GetAll() {
+				unknownNodes[hash] = true
+			}
+			err = s.requestMPTNodes(unknownNodes)
+			if err != nil {
+				// TODO: implement retransmission, it will allow to ignore this error
+				s.log.Warn("failed to request MPT nodes",
+					zap.Uint32("height", s.syncPoint),
+					zap.String("state root", header.PrevStateRoot.StringBE()),
+					zap.Error(err))
+				return err
+			}
+		}
+	}
+
+	if s.syncStage == headersSynced|blocksSynced|mptSynced {
```

Comment:
checkSyncIsCompleted?

Reply:
We don't need to perform a jump in this case because the state is already proper. Thus only check whether all the things were collected and write a log.